src/crowdin_gui.cpp

```diff
@@ -602,20 +592,21 @@ void CrowdinSyncFile(wxWindow *parent, std::shared_ptr<Catalog> catalog,
     // TODO: nicer API for this.
     // This must be done right after entering the modal loop (on non-OSX)
     dlg->CallAfter([=]{
+        const XLIFF1Catalog* xliff = dynamic_cast<const XLIFF1Catalog*>(catalog.get());
+        
         CrowdinClient::Get().UploadFile(
-                str::to_utf8(crowdin_prj), str::to_wstring(crowdin_file), crowdin_lang,
+                xliff->GetProjectId(), xliff->GetFileId(), catalog->GetLanguage().LanguageTag(),
                 catalog->SaveToBuffer()
             )
             .then([=]{
                 auto tmpdir = std::make_shared<TempDirectory>();
-                auto outfile = tmpdir->CreateFileName("crowdin.po");
+                auto outfile = tmpdir->CreateFileName("crowdin.xliff");
```

Comment:
If you changed Crowdin integration to unconditionally use XLIFF (am I understanding that correctly?), how is this going to work with PO files that have plural form entries, given that XLIFF has no support for that?

I think it does make sense to continue using PO download/upload for PO files because of that, for best editing experience, but maybe I'm missing something here?

Reply:
> If you changed Crowdin integration to unconditionally use XLIFF (am I understanding that correctly?)

Yes, correctly

> how is this going to work with PO files that have plural form entries, given that XLIFF has no support for that?

Yes, I just missed about `plurals` so did not take care about this properly, sorry. And this problem definitely need to be addressed in Poedit. Though accordingly to what I figured out with Crowdin team (already as reaction on this your comment) - Crowdin marks plurals in exported Xliff (if they are present in original file, be it `.po` or any other that supports plurals as well, e.g. localization files for Android, iOS, Qt, .NET etc.) in its own way as below
```
<group restype="x-gettext-plurals">
        <trans-unit id="524228[0]" identifier="fd5d778b997859ecdec3298ae0aa67c5">
          <source>one apple</source>
          <target state="needs-translation">one apple</target>
        </trans-unit>
        <trans-unit id="524228[1]" identifier="fd5d778b997859ecdec3298ae0aa67c5">
          <source>other apples  apples appples</source>
          <target state="needs-translation">other apples  apples appples</target>
        </trans-unit>
        <trans-unit id="524228[2]" identifier="fd5d778b997859ecdec3298ae0aa67c5">
          <source>other apples  apples appples</source>
          <target state="needs-translation">other apples  apples appples</target>
        </trans-unit>
        <trans-unit id="524228[3]" identifier="fd5d778b997859ecdec3298ae0aa67c5">
          <source>other apples  apples appples</source>
          <target state="needs-translation">other apples  apples appples</target>
        </trans-unit>
        <note>Context: This text has no context info. The text is used in simple_gettext_plurals.po. Position in file: 1</note>
      </group>
```
i.e. by indicating plurals by `restype="x-gettext-plurals"` and by appending number in square brackets to `id` of `trans-unit` where e.g. `524228[0]` is `id` of base string and `524228[1]`, `524228[2]`, `524228[3]` are `id`s of plurals. The same is handled by Crowdin when importing from `Xliff` as well.
So my proposal that I would implement in Poedit handling of plurals in `Xliff` in the same way as Crowdin does above to be consistent with it while to re-use existing UI/logic/behaviour code in Poedit for displaying/editing them exactly same way it does for `.po`. From my brief look it's just a point of adding implementation of this Crowdin's "hack/workaround/extension" to `Xliff` format somewhere in [cataog_xliff.cpp](https://github.com/tetatetit/poedit/blob/crowdin-new-oauth-api-xliff/src/catalog_xliff.cpp) (still need to figure out where exactly and in most proper way though you are welcome for any hints/advices here just to speed up it for me). This would address plurals **not only for `.po`** but **for all other formats** supported by Crowdin which support plurals as well. So would eliminate necessity to
> continue using PO download/upload for PO files
unless I missing some other advantages of `.po` format here.
So this would fully realize the main idea of Crowdin team to use `Xliff` as the only exchange format between Poedit and Crowdin what would allow Poedit to `Open from Crowdin` **files in any formats** supported by Crowdin for translation. And now with plurals as well after above implementation.

Reply:
In general the advantage of using the native format of the original data is that there is no loss of information. In particular the advantage of using PO format for actual PO files in Poedit is that PO files are best supported by it. 

While supporting `x-gettext-plurals` extension (it is semi-official for gettext conversion into XLIFF, not Crowdin's hack — those are elsewhere) is a good goal in itself, it is extra work and is still not going to be a perfect solution. For example, it doesn't account for plural forms expression, so Poedit would have to use some default — that's fine, except that Crowdin made some interesting choices in that regard in the past (using more forms than necessary due to naive extraction from CLDR data). So that too gets tricky fast.

After briefly checking some XLIFF exports from Crowdin, I see some other issues:

- The exports loose file references. Instead of using XLIFF version of the same, they are thrown into a free-form `<note>Context: ` bit where Crowdin dumps everything it doesn't know how to handle.

- …such as `c-format` and the like (real-world example of the free-form dump: `<note>Context: #. TRANSLATORS: %s is a cloud destination, e.g. &quot;Crowdin&quot; or ftp.wordpress.com etc.
#, c-format</note>`). While fully supported in PO, Poedit would have to have extra parsing just for Crowdin's particular hack of a `<note>` dump.

Reply:
Addressed in #633 704a33e97110dea971574ad4482bff227e618802 seems exactly as discussed here and something by email and and as described/mentioned in above PR and commit message